#!/usr/bin/env cwl-runner
cwlVersion: v1.0
class: CommandLineTool
label: 'Gatk4: SamToFastq'
doc: |2

  USAGE: SamToFastq [arguments]

  Converts a SAM or BAM file to FASTQ. Extracts read sequences and qualities from the input SAM/BAM file and writes them intothe output file in Sanger FASTQ format. See MAQ FASTQ specification for details. This tool can be used by way of a pipe to run BWA MEM on unmapped BAM (uBAM) files efficiently.

  In the RC mode (default is True), if the read is aligned and the alignment is to the reverse strand on the genome, the read sequence from input sam file will be reverse-complemented prior to writing it to FASTQ in order restore correctlythe original read sequence as it was generated by the sequencer.

  Example:

  .. code-tool: none

     java -jar picard.jar SamToFastq \
         I=input.bam \
         FASTQ=output.fastq \

requirements:
- class: ShellCommandRequirement
- class: InlineJavascriptRequirement
- class: DockerRequirement
  dockerPull: |-
    broadinstitute/gatk@sha256:cec850f20311f0686fcf88510bc44e529590d78bec7076a603132115943c09e6

inputs:
- id: javaOptions
  label: javaOptions
  type:
  - type: array
    items: string
  - 'null'
- id: compression_level
  label: compression_level
  doc: |-
    Compression level for all compressed files created (e.g. BAM and VCF). Default value: 2.
  type:
  - int
  - 'null'
- id: inp
  label: inp
  doc: (-I) Input file SAM/BAM to extract reads from. Required.
  type: File
  inputBinding:
    prefix: --INPUT
    separate: true
- id: fastq
  label: fastq
  doc: |-
    (-F) Output FASTQ file (single-end fastq or, if paired first end of the pair FASTQ). Required.
  type:
  - string
  - 'null'
  default: generated.R1.fastq.gz
  inputBinding:
    prefix: --FASTQ
    valueFrom: $(inputs.inp.basename.replace(/.bam$/, "")).R1.fastq.gz
    separate: true
- id: arguments_file
  label: arguments_file
  doc: read one or more arguments files and add them to the command line
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --arguments_file
    separate: true
- id: clipping_action
  label: clipping_action
  doc: |-
    (-CLIP_ACT)  The action that should be taken with clipped reads: 'X' means the reads and qualities should be trimmed at the clipped position; 'N' means the bases should be changed to Ns in the clipped region; and any integer means that the base qualities should be set to that value in the clipped region.
  type:
  - string
  - 'null'
  inputBinding:
    prefix: --CLIPPING_ACTION
    separate: true
- id: clipping_attribute
  label: clipping_attribute
  doc: |-
    (-CLIP_ATTR)  The attribute that stores the position at which the SAM record should be clipped
  type:
  - string
  - 'null'
  inputBinding:
    prefix: --CLIPPING_ATTRIBUTE
    separate: true
- id: clipping_min_length
  label: clipping_min_length
  doc: |-
    (-CLIP_MIN)  When performing clipping with the CLIPPING_ATTRIBUTE and CLIPPING_ACTION parameters, ensure that the resulting reads after clipping are at least CLIPPING_MIN_LENGTH bases long. If the original read is shorter than CLIPPING_MIN_LENGTH then the original read length will be maintained.  Default value: 0.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --CLIPPING_MIN_LENGTH
    separate: true
- id: compress_outputs_per_rg
  label: compress_outputs_per_rg
  doc: |-
    (-GZOPRG)  Compress output FASTQ files per read group using gzip and append a .gz extension to the file names. Default value: false. Possible values: {true, false}
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --COMPRESS_OUTPUTS_PER_RG
    separate: true
- id: create_index
  label: create_index
  doc: |-
    Whether to create an index when writing VCF or coordinate sorted BAM output. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --CREATE_INDEX
    separate: true
- id: create_md5_file
  label: create_md5_file
  doc: |-
    Whether to create an MD5 digest for any BAM or FASTQ files created. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --CREATE_MD5_FILE
    separate: true
- id: ga4gh_client_secrets
  label: ga4gh_client_secrets
  doc: |-
    Google Genomics API client_secrets.json file path.  Default value: client_secrets.json.
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --GA4GH_CLIENT_SECRETS
    separate: true
- id: help
  label: help
  doc: |-
    (-h) display the help message Default value: false. Possible values: {true, false}
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --help
    separate: true
- id: include_non_pf_reads
  label: include_non_pf_reads
  doc: |-
    (-NON_PF)  Include non-PF reads from the SAM file into the output FASTQ files. PF means 'passes filtering'. Reads whose 'not passing quality controls' flag is set are non-PF reads. See GATK Dictionary for more info. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --INCLUDE_NON_PF_READS
    separate: true
- id: include_non_primary_alignments
  label: include_non_primary_alignments
  doc: |-
    If true, include non-primary alignments in the output. Support of non-primary alignments in SamToFastq is not comprehensive, so there may be exceptions if this is set to true and there are paired reads with non-primary alignments. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --INCLUDE_NON_PRIMARY_ALIGNMENTS
    separate: true
- id: interleave
  label: interleave
  doc: |-
    (-INTER) Will generate an interleaved fastq if paired, each line will have /1 or /2 to describe which end it came from. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --INTERLEAVE
    separate: true
- id: max_records_in_ram
  label: max_records_in_ram
  doc: |-
    When writing files that need to be sorted, this will specify the number of records stored in RAM before spilling to disk. Increasing this number reduces the number of file handles needed to sort the file, and increases the amount of RAM needed.  Default value: 500000. 
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --MAX_RECORDS_IN_RAM
    separate: true
- id: output_dir
  label: output_dir
  doc: |-
    Directory in which to output the FASTQ file(s). Used only when OUTPUT_PER_RG is true.
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --OUTPUT_DIR
    separate: true
- id: output_per_rg
  label: output_per_rg
  doc: |-
    (-OPRG) Output a FASTQ file per read group (two FASTQ files per read group if the group is paired). Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --OUTPUT_PER_RG
    separate: true
- id: quality
  label: quality
  doc: |-
    (-Q)  End-trim reads using the phred/bwa quality trimming algorithm and this quality.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --QUALITY
    separate: true
- id: quiet
  label: quiet
  doc: |-
    Whether to suppress job-summary info on System.err. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --QUIET
    separate: true
- id: re_reverse
  label: re_reverse
  doc: |-
    (-RC)  Re-reverse bases and qualities of reads with negative strand flag set before writing them to FASTQ.  Default value: true. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --RE_REVERSE
    separate: true
- id: read1_max_bases_to_write
  label: read1_max_bases_to_write
  doc: |-
    (-R1_MAX_BASES)  The maximum number of bases to write from read 1 after trimming. If there are fewer than this many bases left after trimming, all will be written. If this value is null then all bases left after trimming will be written.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --READ1_MAX_BASES_TO_WRITE
    separate: true
- id: read1_trim
  label: read1_trim
  doc: |-
    (-R1_TRIM)  The number of bases to trim from the beginning of read 1.  Default value: 0.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --READ1_TRIM
    separate: true
- id: read2_max_bases_to_write
  label: read2_max_bases_to_write
  doc: |-
    (-R2_MAX_BASES)  The maximum number of bases to write from read 2 after trimming. If there are fewer than this many bases left after trimming, all will be written. If this value is null then all bases left after trimming will be written.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --READ2_MAX_BASES_TO_WRITE
    separate: true
- id: read2_trim
  label: read2_trim
  doc: |-
    (-R2_TRIM)  The number of bases to trim from the beginning of read 2.  Default value: 0.
  type:
  - int
  - 'null'
  inputBinding:
    prefix: --READ2_TRIM
    separate: true
- id: reference_sequence
  label: reference_sequence
  doc: (-R)  Reference sequence file
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --REFERENCE_SEQUENCE
    separate: true
- id: rg_tag
  label: rg_tag
  doc: |-
    (-RGT)  The read group tag (PU or ID) to be used to output a FASTQ file per read group. Default value: PU
  type:
  - string
  - 'null'
  inputBinding:
    prefix: --RG_TAG
    separate: true
- id: second_end_fastq
  label: second_end_fastq
  doc: (-F2)  Output FASTQ file (if paired, second end of the pair FASTQ).
  type:
  - string
  - 'null'
  default: generated.R2.fastq.gz
  inputBinding:
    prefix: --SECOND_END_FASTQ
    valueFrom: $(inputs.inp.basename.replace(/.bam$/, "")).R2.fastq.gz
    separate: true
- id: showhidden
  label: showhidden
  doc: |-
    (-showHidden)  display hidden arguments.  Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --showHidden
    separate: true
- id: tmp_dir
  label: tmp_dir
  doc: |-
    One or more directories with space available to be used by this program for temporary storage of working files.  This argument may be specified 0 or more times. Default value: null. 
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --TMP_DIR
    separate: true
- id: unpaired_fastq
  label: unpaired_fastq
  doc: |-
    (-FU)  Output FASTQ file for unpaired reads; may only be provided in paired-FASTQ mode.
  type:
  - string
  - 'null'
  default: generated.U.fastq.gz
  inputBinding:
    prefix: --UNPAIRED_FASTQ
    valueFrom: $(inputs.inp.basename.replace(/.bam$/, "")).U.fastq.gz
    separate: true
- id: use_jdk_deflater
  label: use_jdk_deflater
  doc: |-
    (-use_jdk_deflater)  Use the JDK Deflater instead of the Intel Deflater for writing compressed output.  Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --USE_JDK_DEFLATER
    separate: true
- id: use_jdk_inflater
  label: use_jdk_inflater
  doc: |-
    (-use_jdk_inflater)  Use the JDK Inflater instead of the Intel Inflater for reading compressed input.  Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --USE_JDK_INFLATER
    separate: true
- id: validation_stringency
  label: validation_stringency
  doc: |2-
     Validation stringency for all SAM files read by this program.  Setting stringency to SILENT can improve performance when processing a BAM file in which variable-length data (read, qualities, tags) do not otherwise need to be decoded.  Default value: STRICT. Possible values: {STRICT, LENIENT, SILENT} 
  type:
  - string
  - 'null'
  inputBinding:
    prefix: --VALIDATION_STRINGENCY
    separate: true
- id: verbosity
  label: verbosity
  doc: |-
    Control verbosity of logging. Default value: INFO. Possible values: {ERROR, WARNING, INFO, DEBUG} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --VERBOSITY
    separate: true
- id: version
  label: version
  doc: |-
    display the version number for this tool. Default value: false. Possible values: {true, false} 
  type:
  - boolean
  - 'null'
  inputBinding:
    prefix: --version
    separate: true

outputs:
- id: out
  label: out
  doc: |-
    Fastq gzipped output file(s).  May contain single-end or paired-end, or interleaved paired-end.
  type:
    type: array
    items: File
  outputBinding:
    glob: '*.fastq.gz'
    loadContents: false
stdout: _stdout
stderr: _stderr

baseCommand:
- gatk
- SamToFastq
arguments:
- prefix: --java-options
  position: -1
  valueFrom: |-
    $("-Xmx{memory}G {compression} {otherargs}".replace(/\{memory\}/g, (([inputs.runtime_memory, 4].filter(function (inner) { return inner != null })[0] * 3) / 4)).replace(/\{compression\}/g, (inputs.compression_level != null) ? ("-Dsamjdk.compress_level=" + inputs.compression_level) : "").replace(/\{otherargs\}/g, [inputs.javaOptions, []].filter(function (inner) { return inner != null })[0].join(" ")))

hints:
- class: ToolTimeLimit
  timelimit: |-
    $([inputs.runtime_seconds, 86400].filter(function (inner) { return inner != null })[0])
id: GatkSamToFastq
$schemas:
  - 'http://dublincore.org/2012/06/14/dcterms.rdf'
  - 'http://xmlns.com/foaf/spec/20140114.rdf'
  - 'https://schema.org/docs/schema_org_rdfa.html'
's:author':
  - class: 's:Person'
    's:email': 'mailto:rebecca.louise.evans@gmail.com'
    's:identifier': 'https://orcid.org/0000-0002-4923-0662'
    's:name': Rebecca Evans
's:codeRepository': 'https://github.com/beccyl/classi'
's:dateCreated': '2021-01-12'
